text/3861-iter-macro.md

```diff
@@ -0,0 +1,496 @@
+We think it best to let some of these experiments develop further before attempting to support address-sensitive iterators.
+
+See the discussion under [Future Work][self-borrows-future] for more detail about possible approaches to supporting this in the future.
+
+[self-borrow-examples]: https://hackmd.io/DTSOVR4QRLyvaU1HQiQZvg
+
+For more context, see the discussion notes from the [T-lang meeting on Self-Referential Generators](https://hackmd.io/7O9IyhHvRmqaMd-NS6dYyw).
+
+## Why not a Library?
+
+Several crates such as [genawaiter] attempt to provide generators in stable Rust.
+They generally do this by combining a future with a mutable cell to pass values that are being yielded.
+While these are clever uses of futures, they tend to feel like hacks and thus Rust developers a reluctant to use them for anything serious.
+Even as a macro, having iterator blocks built into Rust would let users feel comfortable putting them into production.
+
+As a built-in macro, `iter!` is able to have a cleaner and more efficient implementation by directly constructing internal compiler data structures to expose functionality the compiler already supports.
+
+[genawaiter]: https://github.com/whatisaphone/genawaiter
+
+## Prefix Yield
+
+This RFC is written to allow `yield` in postfix position (`foo.yield`).
+This is a departure from most programming languages, but it is inline with other Rust postfix operators such as `?` or `.await`.
+
+This design decision is primarily forward looking.
+We would like to support full coroutines (see [Future Work][full-coroutines]), at which point `yield` will be able to return a non-`()` value.
+In this case, there will be cases where it is convenient to write `foo().yield.do_something()` instead of `(yield foo()).do_something()`;
+
+Rather than choose one syntax now and either change it or accept a suboptimal syntax in the future, we propose to adopt the postfix `yield` syntax from the start.
+
+# Prior art
+
+Generators are a common feature in many popular programming languages, including Python, JavaScript, C#, Kotlin, PHP, Ruby, etc.
+
+As with futures and `async`/`await`, the Rust version of generators or iterator closures will be more complex than in other languages due to the lack of a garbage collector, as well as the need to manage lifetimes and movability.
+
+# Future possibilities
+
+There are some fairly obvious extensions to this functionality that we have deferred in the interest in shipping a minimum viable product.
+We discuss these in more detail now.
+
+## `IterFn*` Traits
+
+[iter-fn-traits]: #iterfn-traits
+
+Analogously to the `async Fn*` family of traits, we can gain expressiveness by adding a family of `iter Fn*` traits.
+These would allow iterator closures to create iterators that borrow from the closure.
+
+Concretely, `iter Fn*` traits would allow cases like:
+
+```rust
+iter!(|x: &i32| x.yield);
+```
```

Comment:
We [allow this today](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=2dd7bce18609f0e73118786656ea29fd) if you don't specify the type (I think that's an existing limitation with closures not being generic enough).

What the traits allow is specifying bounds with this kind of signature, e.g.

```rust
fn first<'a>(of: &'a str, f: impl IterFn(&str) yields &str) -> &'a str {
    f(x).next().unwrap()
}
```